File: iosApp/iosApp/Modules/NFTSend ✅/NFTSendWireframe.swift
// Created by web3d4v on 04/08/2022.
// Copyright (c) 2022 Sons Of Crypto.
// SPDX-License-Identifier: MIT

import UIKit
import web3walletcore

final class DefaultNFTSendWireframe {
    private weak var parent: UIViewController?
    private let context: NFTSendWireframeContext
    private let qrCodeScanWireframeFactory: QRCodeScanWireframeFactory
    private let confirmationWireframeFactory: ConfirmationWireframeFactory
    private let alertWireframeFactory: AlertWireframeFactory
    private let networksService: NetworksService
    private let currencyStoreService: CurrencyStoreService
    
    private weak var vc: UIViewController?
    
    init(
        _ parent: UIViewController?,
        context: NFTSendWireframeContext,
        qrCodeScanWireframeFactory: QRCodeScanWireframeFactory,
        confirmationWireframeFactory: ConfirmationWireframeFactory,
        alertWireframeFactory: AlertWireframeFactory,
        networksService: NetworksService,
        currencyStoreService: CurrencyStoreService
    ) {
        self.parent = parent
        self.context = context
        self.qrCodeScanWireframeFactory = qrCodeScanWireframeFactory
        self.confirmationWireframeFactory = confirmationWireframeFactory
        self.alertWireframeFactory = alertWireframeFactory
        self.networksService = networksService
        self.currencyStoreService = currencyStoreService
    }
}

extension DefaultNFTSendWireframe: NFTSendWireframe {
    
    func present() {
        let vc = wireUp()
        parent?.show(vc, sender: self)
    }
    
    func navigate(destination_________ destination: NFTSendWireframeDestination) {
        if destination is NFTSendWireframeDestination.UnderConstructionAlert {
            alertWireframeFactory.make(vc, context: .underConstructionAlert()).present()
        }
        if let input = destination as? NFTSendWireframeDestination.QRCodeScan {
            let context = QRCodeScanWireframeContext(
                type: QRCodeScanWireframeContext.Type_Network(network: context.network),
                handler: onPopWrapped(onCompletion: input.onCompletion)
            )
            qrCodeScanWireframeFactory.make(vc, context: context).present()
        }
        if let input = destination as? NFTSendWireframeDestination.ConfirmSendNFT {
            guard input.context.data.addressFrom != input.context.data.addressTo else {
                return presentSendingToSameAddressAlert()
            }
            confirmationWireframeFactory.make(vc, context: input.context).present()
        }
        if destination is NFTSendWireframeDestination.Dismiss {
            vc?.popOrDismiss()
        }
    }
}

private extension DefaultNFTSendWireframe {
    
    func wireUp() -> UIViewController {
        let interactor = DefaultNFTSendInteractor(
            networksService: networksService,
            currencyStoreService: currencyStoreService
        )
        let vc: NFTSendViewController = UIStoryboard(.nftSend).instantiate()
        let presenter = DefaultNFTSendPresenter(
            view: WeakRef(referred: vc),
            wireframe: self,
            interactor: interactor,
            context: context
        )
        vc.hidesBottomBarWhenPushed = true
        vc.presenter = presenter
        self.vc = vc
        return vc
    }
    
    func presentSendingToSameAddressAlert() {
        alertWireframeFactory.make(
            vc,
            context: .init(
                title: Localized("alert.send.transaction.toYourself.title"),
                media: .image(named: "hand.raised", size: .init(length: 40)),
                message: Localized("alert.send.transaction.toYourself.message"),
                actions: [
                    .init(title: Localized("Ok"), type: .primary, action: nil)
                ],
                contentHeight: 230
            )
        ).present()
    }
    
    func onPopWrapped(onCompletion: @escaping (String) -> Void) -> (String) -> Void {
        {
            [weak self] string in
            _ = self?.vc?.navigationController?.popViewController(animated: true)
            onCompletion(string)
        }
    }
}

File: androidApp/src/main/java/com/sonsofcrypto/web3wallet/android/modules/compose/nftsend/NFTSendWireframe.kt
package com.sonsofcrypto.web3wallet.android.modules.compose.nftsend

import androidx.fragment.app.Fragment
import com.sonsofcrypto.web3lib.services.currencyStore.CurrencyStoreService
import com.sonsofcrypto.web3lib.services.networks.NetworksService
import com.sonsofcrypto.web3lib.utils.WeakRef
import com.sonsofcrypto.web3wallet.android.R
import com.sonsofcrypto.web3wallet.android.common.NavigationFragment
import com.sonsofcrypto.web3wallet.android.common.extensions.navigationFragment
import com.sonsofcrypto.web3wallet.android.modules.compose.confirmation.ConfirmationWireframeFactory
import com.sonsofcrypto.web3wallet.android.modules.compose.qrcodescan.QRCodeScanWireframeFactory
import com.sonsofcrypto.web3walletcore.modules.nftSend.*
import com.sonsofcrypto.web3walletcore.modules.qrCodeScan.QRCodeScanWireframeContext

class DefaultNFTSendWireframe(
    private val parent: WeakRef<Fragment>?,
    private val context: NFTSendWireframeContext,
    private val networksService: NetworksService,
    private val currencyStoreService: CurrencyStoreService,
    private val confirmationWireframeFactory: ConfirmationWireframeFactory,
    private val qrCodeScanWireframeFactory: QRCodeScanWireframeFactory,
): NFTSendWireframe {

    private lateinit var fragment: WeakRef<Fragment>

    override fun present() {
        val fragment = wireUp()
        this.fragment = WeakRef(fragment)
        parent?.get()?.childFragmentManager?.beginTransaction()?.apply {
            add(R.id.container, fragment)
            commitNow()
        }
    }

    override fun navigate(destination: NFTSendWireframeDestination) {
        when (destination) {
            is NFTSendWireframeDestination.UnderConstructionAlert -> {}
            is NFTSendWireframeDestination.QRCodeScan -> {
                val context = QRCodeScanWireframeContext(
                    type = QRCodeScanWireframeContext.Type.Network(network = context.network),
                    handler = onPopWrapped(onCompletion = destination.onCompletion)
                )
                qrCodeScanWireframeFactory.make(parent?.get(), context).present()
            }
            is NFTSendWireframeDestination.ConfirmSendNFT -> {
                confirmationWireframeFactory.make(
                    parent = parent?.get(),
                    context = destination.context
                ).present()
            }
            is NFTSendWireframeDestination.Dismiss -> {
                println("[AA] Implement -> navigateTo $destination")
            }
        }
    }

    private fun wireUp(): Fragment {
        val view = NFTSendFragment()
        val interactor = DefaultNFTSendInteractor(
            networksService,
            currencyStoreService,
        )
        val presenter = DefaultNFTSendPresenter(
            WeakRef(view),
            this,
            interactor,
            context,
        )
        view.presenter = presenter
        return NavigationFragment(view)
    }

    private fun onPopWrapped(onCompletion: (String) -> Unit): (String) -> Unit = { addressTo ->
        parent?.get().navigationFragment?.dismiss(
            onCompletion = { onCompletion(addressTo) }
        )
    }
}